src/Service.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace Dhii\Services;
+
+use Psr\Container\ContainerInterface;
+
+/**
+ * Partial implementation of a service.
+ *
+ * This class represents a DI service in functor form. Instances of this class or its derivations are invocable via
+ * the magic {@link __invoke()} method. The signature of this method is equivalent to that of regular service definition
+ * functions, in that they accept a {@link ContainerInterface} instance.
+ *
+ * Services are also aware of any services that they depend on, by key. These keys may be used by derivations of this
+ * class to achieve some automation, be it during run-time or for static analysis purposes.
+ *
+ * @since [*next-version*]
+ * @see   __invoke()
+ * @see   ContainerInterface
+ */
+abstract class Service
+{
+    /**
+     * @since [*next-version*]
+     *
+     * @var string[]
+     */
+    protected $dependencies;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param string[] $dependencies A list of keys that correspond to other services that this service depends on.
+     */
+    public function __construct(array $dependencies)
+    {
+        $this->dependencies = $dependencies;
+    }
+
+    /**
+     * Retrieves the keys of dependent services.
+     *
+     * @since [*next-version*]
+     *
+     * @return string[] A list of strings each representing the key of a service.
+     */
+    public function getDependencies() : array
+    {
+        return $this->dependencies;
+    }
+
+    /**
+     * Creates a copy of this service with different dependency keys.
+     *
+     * @since [*next-version*]
+     *
+     * @param array $dependencies The new service dependency keys.
+     *
+     * @return static The newly created service instance.
+     */
+    public function withDependencies(array $dependencies)
+    {
+        $instance = clone $this;
+        $instance->dependencies = $dependencies;
+
+        return $instance;
+    }
+
+    /**
+     * Creates a value for this service using a given container for dependency resolution.
+     *
+     * @since [*next-version*]
+     *
+     * @param ContainerInterface $c The container to use to resolve dependencies.
+     *
+     * @return mixed The created service value.
+     */
+    abstract public function __invoke(ContainerInterface $c);
+
+    /**
+     * Resolves a set of service keys using a given container.
+     *
+     * @since [*next-version*]
+     *
+     * @param ContainerInterface $c    The container to use for service resolution.
+     * @param array              $keys The services keys to resolve.
+     *
+     * @return array A list containing the resolved service values, in the same as given by the $keys argument. All
+     *               indices from the $keys argument will be preserved.
+     */
+    public static function resolveKeys(ContainerInterface $c, array $keys)
```

Comment:
Usage appears to be restricted to descendants of this class. If other usage is needed, this is trivial to implement. Perhaps it would be better to make it an instance method rather than `static`, and invoke it in descendants as `parent::resolveKeys()` rather than `Service::resolveKeys()`.

Reply:
I prefer static because it keeps the algorithm decoupled from instances, potentially allowing a non-descendant to call it. Example: `MyStaticAnalysisTool` can also call `Service::resolveKeys()` without re-implementing it. The pattern I'm employing as of recent is to make `public` functionality be methods and everything else be static to promote re-usability.

This adds some development latency, as a dev needs to carefully consider in which class the static functionality should belong to. But I'm willing to live with that given that devs already burden themselves with namespace design, which is basically what this boils down to; classes becoming autoloadable namespaces for functions.

Reply:
Static promotes IMHO the _wrong way_ of re-usability. Cannot inject, and cannot decorate. So, I'd really appreciate if you'd change it so that we can take advantage of those patterns in the near future.

Reply:
Please remember that the aim of this package is to provide a set of service helpers that are uniform across all dependents. Injecting an decorating go against that idea. These implementations are helpers and should not be viewed as objects, but as functors. So I see no reason to change this to be non-static.